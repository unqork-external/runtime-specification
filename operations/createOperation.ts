import type { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'
import type { CustomOperationTypeMap } from '../../operations'
import type { InputRef } from '../inputs/inputRef'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: Partial<AdditionalArgs>,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

export const createCustomOperation = <OperationEnum extends string, OpType extends OperationEnum>(
  type: OpType,
  options: InstanceType<CustomOperationTypeMap<OperationEnum>[OpType]['model']>['options'],
  typeMap: CustomOperationTypeMap<OperationEnum>,
  additionalArgs?: Partial<AdditionalArgs>,
) => {
  const classOpType = typeMap[type].model
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  name: string
  notifyImmediately: boolean
  creatorSummary: string
}
