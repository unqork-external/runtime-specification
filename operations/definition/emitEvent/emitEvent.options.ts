import { Description, Example, Required, Optional } from '@tsed/schema'

import { OperationOptions } from '../../interface'

export class EmitEventOperationOptions extends OperationOptions {
  @Required()
  @Example('myCustomEvent')
  @Description('a creator defined name of an event to fire in the system')
  event: string

  @Optional()
  @Example('{value: "this is a value"}')
  @Description('data that can be sent with the event')
  payload: Record<string, any>
}
