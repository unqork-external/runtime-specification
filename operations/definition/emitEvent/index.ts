export * from './emitEvent.model'
export * from './emitEvent.options'
