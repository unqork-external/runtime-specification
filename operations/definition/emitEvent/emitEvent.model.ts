import { DiscriminatorValue, Required } from '@tsed/schema'

import { EmitEventOperationOptions } from './emitEvent.options'
import { Alpha, DisplayName } from '../../../../decorators'
import { AutogenDisabled } from '../../decorators/autogenDisabled.decorator'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DisplayName('Emit Event')
@DiscriminatorValue(OperationTypes.EMIT_EVENT)
@Alpha()
@AutogenDisabled()
export class EmitEventOperation extends Operation<OperationTypes.EMIT_EVENT> {
  public type = OperationTypes.EMIT_EVENT as const

  @Required()
  public options: EmitEventOperationOptions = new EmitEventOperationOptions()
}
