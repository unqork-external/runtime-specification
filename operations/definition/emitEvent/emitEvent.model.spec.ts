import { EmitEventOperation } from './emitEvent.model'
import { generateSchemaAndValidate } from '../../../../utilities'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('Operation Model: EmitEvent Tests', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(EmitEventOperation)
    validate = response.validate
    schema = response.schema
  })

  it('Reset operation should validate appropriately', () => {
    const op = createOperation(OperationTypes.EMIT_EVENT, {
      targetKey: '',
      event: 'helloEvent',
      payload: {},
    })
    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })
})
