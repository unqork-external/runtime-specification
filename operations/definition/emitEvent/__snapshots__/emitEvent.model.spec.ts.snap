// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operation Model: EmitEvent Tests should match this JSON Schema 1`] = `
{
  "definitions": {
    "EmitEventOperationOptions": {
      "properties": {
        "event": {
          "description": "a creator defined name of an event to fire in the system",
          "examples": [
            "myCustomEvent",
          ],
          "minLength": 1,
          "type": "string",
        },
        "inputs": {
          "description": "
    Inputs to be used within formulas in other operation options.
  ",
          "items": {
            "$ref": "#/definitions/InputRef",
          },
          "type": "array",
        },
        "payload": {
          "description": "data that can be sent with the event",
          "examples": [
            "{value: "this is a value"}",
          ],
          "type": "object",
        },
        "shouldExecute": {
          "default": true,
          "description": "
    Conditionally execute this operation.
    Can set to false to temporarily disabled when debugging.
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "targetKey": {
          "description": "This is not used for this event but is required by all operation options",
          "type": "string",
        },
      },
      "required": [
        "event",
      ],
      "type": "object",
    },
    "InputRef": {
      "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
      "properties": {
        "alias": {
          "type": "string",
        },
        "propertyPath": {
          "type": "string",
        },
        "required": {
          "type": "boolean",
        },
        "silent": {
          "type": "boolean",
        },
        "targetKey": {
          "minLength": 1,
          "type": "string",
        },
        "type": {
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/EmitEventOperationOptions",
    },
    "type": {
      "const": "EMIT_EVENT",
      "examples": [
        "EMIT_EVENT",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
